### PR TITLE
feat(avoidance): make it possible to output small shift avoidance path

### PR DESCRIPTION
**Pull request by satoshi-ota**
## Description

Improve shift line pre-process logic so that the module outputs small shift avoidance path. 

- Grid: 0.5 x 0.5 m
- Green Path: original path
- Red Path: candidate path

image

This PR includes following modifications:

- add fill gap process before the merge process in order to remove noisy shift line.
- add one more small longitudinal shift line trimmer in order to remove noisy shift line.
- fix shift line start/end shift length based on previous linear shift line to remove noisy shift line.
- check shift gradient sign (positive or negative) in similar gradient shift line merger function in order to prevent vanishing of small shift line.

(Before this PR) huge gaps exist in merged shift lines.

image

(After this PR) merged shift lines are connected smoothly.

image

## Tests performed

- [x] [PASS TIER IV INTERNAL SCENARIOS](https://evaluation.tier4.jp/evaluation/reports/5444c132-eaf9-529f-831e-9df3f9c109dd?project_id=prd_jt)

## Effects on system behavior

Make it possible to output small shift avoidance path.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 1]
```diff
@@ -1181,22 +1214,81 @@ AvoidLineArray AvoidanceModule::extractShiftLinesFromLine(ShiftLineData & shift_
...
+      new_line.start_shift_length = helper_.getEgoLinearShift();
+      new_line.start_longitudinal = 0.0;
+      new_line.end_shift_length = front.start_shift_length;
+      new_line.end_longitudinal = front.start_longitudinal;
+      new_line.id = getOriginalShiftLineUniqueId();
+
+      output.push_back(new_line);
+    }
+  }
+
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1267]: how about using lambda function?
```suggestion
    auto createNewAvoidLine = [this](double start_shift_length, double start_longitudinal,
                                    double end_shift_length, double end_longitudinal) {
      AvoidLine newLine{};
      newLine.start_shift_length = start_shift_length;
      newLine.start_longitudinal = start_longitudinal;
      newLine.end_shift_length = end_shift_length;
      newLine.end_longitudinal = end_longitudinal;
      newLine.id = getOriginalShiftLineUniqueId();
      return new_line;
    };

    if (has_gap) {
      auto new_line = createNewAvoidLine(helper_.getEgoLinearShift(), 0.0, firstLine.start_shift_length, 
                                      firstLine.start_longitudinal);
      output.push_back(new_line);
    }
  }

  for (size_t i = 0; i < sorted_input.size() - 1; ++i) {
    const auto front = sorted_input.at(i);
    const auto back = sorted_input.at(i + 1);
    const auto has_gap = back.start_longitudinal - front.end_longitudinal > 0.0;

    if (has_gap) {
      auto new_line = createNewAvoidLine(currentLine.end_shift_length, currentLine.end_longitudinal, 
                                        nextLine.start_shift_length, nextLine.start_longitudinal);
```

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 2]
```diff
@@ -1112,18 +1129,35 @@ void AvoidanceModule::generateTotalShiftLine(
...
+  using utils::avoidance::setStartData;
+
   const auto & path = avoidance_data_.reference_path;
   const auto & arcs = avoidance_data_.arclength_from_ego;
   const auto N = path.points.size();
 
   auto & sl = shift_line_data;
 
-  const auto getBwdGrad = [&](const size_t i) {
+  const auto backward_grad = [&](const size_t i) {
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1160]: Here is also the point you can use lambda function if you think it's good.
```suggestion
  auto calculate_gradient = [&](const size_t i, const bool is_forward) {
    const int direction = is_forward ? 1 : -1;
    if (i == 0 && !is_forward || i == arcs.size() - 1 && is_forward) {
      return sl.shift_line_grad.at(i);
    }

    const double ds = arcs.at(i) - arcs.at(i - direction);
    if (ds < 1.0e-5) {
      return sl.shift_line_grad.at(i);
    }

    return (sl.shift_line.at(i) - sl.shift_line.at(i - direction)) / ds;
  };

```

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 3]
```diff
@@ -1166,8 +1200,7 @@ AvoidLineArray AvoidanceModule::extractShiftLinesFromLine(ShiftLineData & shift_
     // If the vehicle is already on the avoidance (checked by the first point has shift),
     // set a start point at the first path point.
     if (!found_first_start && std::abs(shift) > IS_ALREADY_SHIFTING_THR) {
-      utils::avoidance::setStartData(
-        al, 0.0, p, i, arcs.at(i));  // start length is overwritten later.
+      setStartData(al, 0.0, p, i, arcs.at(i));  // start length is overwritten later.
       found_first_start = true;
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1204]: `found_first_start` is whether start point of avoid line is found or not, correct?

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 2 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1160]: Thanks for your advice!, But I think it is better to separate backward/forward function for readability. I'll fix it in other PR if necessary :+1:

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 3 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1263]: You're right!

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 1 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1330]: Use lambda in https://github.com/autowarefoundation/autoware.universe/pull/3852/commits/96af6872d7bea07db5632efa38670c54252a683f.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 4]
```diff
@@ -1893,6 +1979,10 @@ void AvoidanceModule::addReturnShiftLineFromEgo(
       }
     }
 
+    if (has_candidate_point || has_registered_point) {
```
[the remark this answers is not in the document]
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2044]: Add brief comment in https://github.com/autowarefoundation/autoware.universe/pull/3852/commits/96af6872d7bea07db5632efa38670c54252a683f.

**Pull request merged**